packages/pyolite-kernel/py/pyolite/pyolite/kernel.py

```diff
@@ -15,3 +17,30 @@ def comm_info(self, target_name=""):
                 comms[comm_id] = dict(target_name=comm.target_name)
 
         return comms
+
+    def do_complete(self, code, cursor_pos):
+        if cursor_pos is None:
+            cursor_pos = len(code)
+        line, offset = line_at_cursor(code, cursor_pos)
+        line_cursor = cursor_pos - offset
+
+        txt, matches = self.interpreter.complete("", line, line_cursor)
+        return {
+            "matches": matches,
+            "cursor_end": cursor_pos,
+            "cursor_start": cursor_pos - len(txt),
+            "metadata": {},
+            "status": "ok",
+        }
+
+    async def run(self, code):
+        self.interpreter._last_traceback = None
+        exec_code = self.interpreter.transform_cell(code)
+        await _load_packages_from_imports(exec_code)
+        if self.interpreter.should_run_async(code):
+            self.result = await self.interpreter.run_cell_async(
+                code, store_history=True
+            )
+        else:
+            self.result = self.interpreter.run_cell(code, store_history=True)
+        return self.result
```

Comment:
Your `do_complete` returns a message that respects the kernel protocol, maybe we could move some logic that you have in the `worker.ts` here, to return the proper protocol message from here in Python.

Reply:
so instead of constructing a `reply` on the JS side, we return the `reply` on the python side?

Reply:
Yes. Well, it's just that because you do it for the `complete`, it seems more homogeneous to me to do it for the `run` as well. Also this makes the code closer to ipykernel, and could potentially help us to refactor to share more code.